### ISSUE TITLE
New verb for waiting for change outside RE control

### ISSUE THREAD

**Issue by DominicOram** [bluesky/bluesky#1618]
There a number of cases where there is something changing outside of RE control e.g. typically the beam, or maybe something that is being driven by an IOC internally/external process etc. It is currently hard to write the ophyd device/bluesky plan to wait on this change as the framework assumes you're waiting on the results of a `set`/`trigger`, neither of which quite match this case. (as discussed with @danielballan during code review)

## Expected Behavior
With an ophyd device something like:
```python
class Beam(Device):
    current: EpicsSignal = Component(EpicsSignal, "CURRENT")
    def read_wait():
         return SubscriptionStatus(current, lambda *, old_value, value, **kwargs: value > 300 , timeout=timeout)
```

```python
def my_plan():
    yield from bps.read_wait(beam)
```

Note that I'm not sure on the name of the verb, `wait` would seem reasonable but already exists.

## Current Behavior
There are ways you can do this:
* Use `set`/`trigger` despite nothing is being set/triggered - I would say this has the potential to confuse
* Use a polling loop in the RE like [ramp](https://github.com/bluesky/bluesky/blob/6e3c4721ce728bcea2d2cc64f3e56d7c6b31e5e5/bluesky/plans.py#L1863) - This is a bit verbose and there's no standard way of doing it

**Comment by dmgav** [comment]
I think it would be useful to have a explicit method for waiting for a PV with timeout both in Ophyd and Bluesky.

**Comment by prjemian** [comment]
Why not wait for a status object that is finished by custom code run in a
thread?  Timeout is part of that code. Details are different, depending on
the PV and what is expected.

I don't see a need to expand the language, just a demo example.

**Comment by untzag** [comment]
Would this change be in addition to #1596? Would it obviate #1596?

**Comment by coretl** [comment]
How about:
```python
from ophyd_async.core import wait_for_value

def my_plan():
    coro = wait_for_value(beam_current, lambda v: v > 300, timeout=60)
    yield from bps.wait_for([coro])
```

**Comment by coretl** [comment]
Wait, the docs for `bps.wait_for` are wrong, it takes a bunch of future factories, not awaitables, so it would be:
```python
def my_plan():
    yield from bps.wait_for(
        [lambda: wait_for_value(beam_current, lambda v: v > 300, timeout=60)]
    )
```

**Comment by callumforrester** [comment]
@untzag Possibly something like #1596 could serve as the underlying verb for lots of useful plan stubs such as `wait_for_value`, the difference is that #1596 generates a status object under the hood whereas @coretl's example just passes a coroutine to the `RunEngine` and makes it `await` it. Is there any particular advantage to having the status object, @tacaswell?

**Comment by tacaswell** [comment]
I think the advantage of doing this with a status object is that you can make it the problem of the ophyd object to decide when the condition is met rather than that logic living in the plan.

It seems more consistent with the overall design, we have: "tell me when the trigger is done", "tell me when the fly step is done", and "tell me when the 'motion' is done" as first-class verbs/concepts.  Adding "tell me when the outside world has pushed you to a particular state" completes the set (and would make boot-strapping any future specialized verbs easier as each of those is "do something" then "tell me when world meets the state I asked for").

I suspect that in 90% of cases, wait_for_value will be good enough, but we should not engineer our selves out of that last 10% (the thing I'm thinking of now is multi-signal conditions).

**Comment by tacaswell** [comment]
Using a status object also gives us the hooks to add progress bars which make make sense in some cases.

**Comment by coretl** [comment]
If we either expand `Msg("wait")` to take `statuses` instead of `group` (inspiration from a conversation with @danielballan today) or make a new `Msg("wait_for_statuses")` then we could (after https://github.com/bluesky/ophyd-async/pull/176):
```python
@WatchableAsyncStatus.wrap
def wait_for_signal_above(signal, threshold, timeout):
    initial = await signal.get_value()
    async for current in observe_value(signal, total_timeout=timeout):
        yield WatcherUpdate(current=current, target=threshold, initial=initial)
        if current > threshold:
            break

def my_plan():
    yield from bps.wait_for_statuses(wait_for_signal_above(beam_current_signal, 300, timeout=60))
```

**Comment by tacaswell** [comment]
Another major advantage of adding and ophyd verb is that we can allow the RE to create the status objects rather than creating them in the plan so if plans are run through other consumers we won't have a bunch of dangling status objects / run awaited co-routines.

**Comment by coretl** [comment]
> Another major advantage of adding and ophyd verb is that we can allow the RE to create the status objects rather than creating them in the plan so if plans are run through other consumers we won't have a bunch of dangling status objects / run awaited co-routines.

Could we make a new device and reuse a verb on it? E.g. `trigger`?

```python
class BeamCurrentWaitable(Device):
    def __init__(self, name=None):
        self.beam_current = epics_signal_r("BEAM:CURRENT")
        super().__init__(name)

    @WatchableAsyncStatus.wrap
    def trigger(self):
        initial = await self.beam_current.get_value()
        async for current in observe_value(self.beam_current):
            yield WatcherUpdate(current=current, target=threshold, initial=initial)
            if current > 300:
                break

def my_plan():
    yield from bps.trigger(beam_current_waitable)
```

**Comment by DominicOram** [comment]
> Could we make a new device and reuse a verb on it? E.g. trigger?

See my original comment about this being confusing as nothing is actually being triggered.

Unless I'm mistaken all your suggestions above rely on `ophyd-async` @coretl? Would be nice to have an `ophyd`/`ophyd-async` agnostic solution, though not essential for me as we are moving everything to `ophyd-async` anyway.

**Comment by coretl** [comment]
> See my original comment about this being confusing as nothing is actually being triggered.

Good point... In which case `s/trigger/some_new_verb` in the example above...

> Unless I'm mistaken all your suggestions above rely on `ophyd-async` @coretl? Would be nice to have an `ophyd`/`ophyd-async` agnostic solution, though not essential for me as we are moving everything to `ophyd-async` anyway.

I think the idea of "new ophyd verb that makes a status", and "new run engine verb that calls it" is reasonable. Threading the status class through the RE makes me nervous as I couldn't then do something like the example above. Although making it a status factory would probably be all that is needed as then we could make a utility `def wait_for_signal_to_exceed(signal, threshold) -> WatchableAsyncStatus`.

The implementations I've suggested above is how I would implement it in ophyd-async (for the more complicated example of watcher updates to a status), there would be nothing to stop it being implemented in ophyd in a similar way.

**Comment by tacaswell** [comment]
We have some examples of doing this on the floor at NSLS-II, but we hi-jack `set` (because `set` seemed closer than `trigger` to my brain).

I think the consenus here is to do this and it may be good for the hackathon next week.  I think the exact steps are:

 1. pick a name (90% of the effort)
 2. add a protocol for it
 3. teach the RE the new verb and allow it to participate in the "group" tracking
 4. add reference implementation in ohpyd-async
 5. add reference implementation in ophyd
 6. document, document, document
 7. bonus points for refactoring set/trigger/kickoff/complete to use the same (internal) implementation if practical

**Comment by callumforrester** [comment]
Reviewers: everyone present
